### PR TITLE
Ensure elision of `require_one_based_indexing` with high-dim array views 

### LINKED ISSUE #998: Regression: `mul!` of views allocates in v1.9+ for arrays of 4 or higher dimension, not in v1.8.5
This seems like a regression: in Julia v1.9-rc2 and master
```julia
julia> using LinearAlgebra, BenchmarkTools

julia> A = rand(ComplexF64,4,4,1000,1000);

julia> B = similar(A);

julia> a,b = (view(B,:,:,1,1),view(A,:,:,1,1));

julia> @btime mul!($b,$a,$a);
  321.792 ns (10 allocations: 608 bytes)
```
This is on an M1 Max
```julia
julia> versioninfo()
Julia Version 1.9.0-rc2
Commit 72aec423c2a (2023-04-01 10:41 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin21.3.0)
  CPU: 10 × Apple M1 Max
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, apple-m1)
  Threads: 1 on 8 virtual cores
Environment:
  JULIA_EDITOR = code
```

But in v1.8.5 we don't allocate
```julia
julia> using LinearAlgebra, BenchmarkTools

julia> A = rand(ComplexF64,4,4,1000,1000);

julia> B = similar(A);

julia> a,b = (view(B,:,:,1,1),view(A,:,:,1,1));

julia> @btime mul!($b,$a,$a);
  97.076 ns (0 allocations: 0 bytes)
```
Note that if A has dimensions 3 or less, this issue does not arise.

### LINKED ISSUE #49332: Regression: `mul!` of views allocates in v1.9+ for arrays of 4 or higher dimension, not in v1.8.5
This seems like a regression: in Julia v1.9-rc2 and master
```julia
julia> using LinearAlgebra, BenchmarkTools

julia> A = rand(ComplexF64,4,4,1000,1000);

julia> B = similar(A);

julia> a,b = (view(B,:,:,1,1),view(A,:,:,1,1));

julia> @btime mul!($b,$a,$a);
  321.792 ns (10 allocations: 608 bytes)
```
This is on an M1 Max
```julia
julia> versioninfo()
Julia Version 1.9.0-rc2
Commit 72aec423c2a (2023-04-01 10:41 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin21.3.0)
  CPU: 10 × Apple M1 Max
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, apple-m1)
  Threads: 1 on 8 virtual cores
Environment:
  JULIA_EDITOR = code
```

But in v1.8.5 we don't allocate
```julia
julia> using LinearAlgebra, BenchmarkTools

julia> A = rand(ComplexF64,4,4,1000,1000);

julia> B = similar(A);

julia> a,b = (view(B,:,:,1,1),view(A,:,:,1,1));

julia> @btime mul!($b,$a,$a);
  97.076 ns (0 allocations: 0 bytes)
```
Note that if A has dimensions 3 or less, this issue does not arise.

### PR DESCRIPTION
**Pull request by pablosanjose**
Closes JuliaLang/LinearAlgebra.jl#998

Here I try to make the life of the compiler a bit easier by reorganising the dispatch structure of `has_offset_axes` (#45260) into two layers. Avoiding the self-reference of the `has_offset_axes(As...)` method seems to allow the elision of the call in my machine, which fixes JuliaLang/LinearAlgebra.jl#998 by fully eliminating `require_one_based_indexing` even for the problematic case of views of 4+ dimensional arrays.

### CODE REVIEW

**Comment by pablosanjose** [comment]
This fixed one case but broke others. Tweaks are needed still

**Comment by pablosanjose** [comment]
The first fix was not the correct one. With the last commit we simplify the lispy recursion of `has_offset_axes(As...)` without using `_any_tuple`. This assumes that `has_offset_axes(A)` is always a Bool, and never Missing. If this assumption holds, the new and old code should be equivalent. Otherwise, if we hit a `has_offset_axes(A)::Missing` both master and PR would  throw a `TypeError: non-boolean (Missing) used in boolean context`, but at different points of the call chain (`require_one_based_indexing` vs `has_offset_axes`).

CC: @N5N3

**Comment by N5N3** [comment]
LGTM, I guess the doc string could be updated? (`|` -> `||`)

**Review by N5N3 — APPROVED** [review]

**Comment by pablosanjose** [comment]
> I guess the doc string could be updated? (| -> ||)

Done, thanks!

File: base/abstractarray.jl [hunk 1]
```diff
@@ -103,17 +103,20 @@ end
...
+has_offset_axes() = false
 has_offset_axes(A) = _any_tuple(x->Int(first(x))::Int != 1, false, axes(A)...)
 has_offset_axes(A::AbstractVector) = Int(firstindex(A))::Int != 1 # improve performance of a common case (ranges)
-# Use `_any_tuple` to avoid unneeded invoke.
-# note: this could call `any` directly if the compiler can infer it
-has_offset_axes(As...) = _any_tuple(has_offset_axes, false, As...)
 has_offset_axes(::Colon) = false
 has_offset_axes(::Array) = false
+# note: this could call `any` directly if the compiler can infer it. We don't use _any_tuple
+# here because it stops full elision in sone cases (#49332) and we don't need handling of
```
Comment (barucden) [inline base/abstractarray.jl:116]:
```suggestion
# here because it stops full elision in some cases (#49332) and we don't need handling of
```
tiny typo

Comment (pablosanjose) [reply base/abstractarray.jl:116]: thanks!

**Pull request merged**